### PR TITLE
feat: validate local and external references

### PR DESCRIPTION
**Pull request by probablyArth**
### validate external references
- external url is invalid
image
- external url leads to invalid json-schema file
image
- relative external url but invalid `$id`
image

### validate internal references
image

TODO:
- validate anchor fragments
- look for optimizations

### CODE REVIEW

**Comment by probablyArth** [comment]
solves #7

**Comment by jdesrosiers** [comment]
Sorry, but that's not what is meant by "external reference". It means external to the schema document, not external to the workspace. I wasn't thinking about fetching schemas over HTTP and at this point and I don't want to go there. There are complications involved in doing that right that we don't want to deal with right now. For this task, you should only be considering schemas defined in the workspace or pre-registered in `@hyperjump/json-schema` like the meta-schemas.

`my-project/schemas/address.schema.json`
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/address",

  ...
}
```

`my-project/schemas/customer.schema.json`
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/customer",

  "type": "object",
  "properties": {
    ...,
    "address": { "$ref": "/schemas/address" } <-- External reference to a schema in the workspace
  }
}
```

**Comment by probablyArth** [comment]
> Sorry, but that's not what is meant by "external reference". It means external to the schema document, not external to the workspace. I wasn't thinking about fetching schemas over HTTP and at this point and I don't want to go there. There are complications involved in doing that right that we don't want to deal with right now. For this task, you should only be considering schemas defined in the workspace or pre-registered in `@hyperjump/json-schema` like the meta-schemas.
> 
> `my-project/schemas/address.schema.json`
> 
> ```json
> {
>   "$schema": "https://json-schema.org/draft/2020-12/schema",
>   "$id": "https://example.com/schemas/address",
> 
>   ...
> }
> ```
> 
> `my-project/schemas/customer.schema.json`
> 
> ```json
> {
>   "$schema": "https://json-schema.org/draft/2020-12/schema",
>   "$id": "https://example.com/schemas/customer",
> 
>   "type": "object",
>   "properties": {
>     ...,
>     "address": { "$ref": "/schemas/address" } <-- External reference to a schema in the workspace
>   }
> }
> ```

Ahh! Got it!
will make the necessary changes. I was also a little skeptical when fetching schemas over http, should've asked beforehand my bad 😅

File: language-server/src/references.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+   * @returns
+   */
+  async function validateRefs(instance, basePath = "") {
+    if (instance.typeOf() === "object") {
+      for (const [key, valueInstance] of instance.entries()) {
+        if (key.value() == "$id" && typeof valueInstance.value() === "string") {
+          baseUri = extractBaseUri(valueInstance.value());
+        }
+        if (
+          key.value() === "$ref" &&
```
Comment (jdesrosiers) [inline language-server/src/references.js:30]: The vocabulary system allows you to do a lot unusual things including changing the name of the token used to represent a reference. That means you can't just hardcode to the string `$ref`. To get correct keyword name for a given dialect, use the [`getKeywordName`](https://github.com/hyperjump-io/json-schema?tab=readme-ov-file#api-2) function from `@hyperjump/json-schema`. There are two versions of references using the keywordIds `https://json-schema.org/keyword/ref` and `https://json-schema.org/keyword/draft-04/ref`. For the purposes of this functionality there isn't a difference between the two, but you'll need to check for both.

File: language-server/src/json-instance-from-json.js [hunk 2]
```diff
@@ -0,0 +1,261 @@
+import { parseTree } from "jsonc-parser";
+import * as JsonPointer from "@hyperjump/json-pointer";
+import { getKeywordId } from "@hyperjump/json-schema/experimental";
+import { drop, find, head, some } from "@hyperjump/pact";
+import { toAbsoluteUri } from "./util.js";
+
+export class JsoncInstanceFromJSON {
```
Comment (jdesrosiers) [inline language-server/src/json-instance-from-json.js:7]: There doesn't look like there's a good reason to create a new Instance class and the duplication is unacceptable. It looks like all you really needed was to add your `fromJSON` static function to the existing `JsoncInstance` class.

File: language-server/src/references.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+  validateRefs(instance);
+  await Promise.all(promises);
+
+  return diagnostics;
+};
+
+function isLocalReference(ref) {
+  return ref.startsWith("#");
+}
```
Comment (jdesrosiers) [inline language-server/src/references.js:72]: This isn't going to be sufficient. What about a reference like, `another-schema#/nothing/here`?

**Review by jdesrosiers — COMMENTED** [review]
Just a quick review to point out a couple things. I review in too much detail yet, because I know you're going to have to change a lot of it to get back on the right track.

File: language-server/src/references.js (hunk 3 above)
Comment (probablyArth) [reply language-server/src/references.js:75]: but `another-schema#/` is not a local reference right?
I couldn't find a example on the documentation which starts without `/` or `#` or a baseUri

**Comment by benodiwal** [comment]
> Sorry, but that's not what is meant by "external reference". It means external to the schema document, not external to the workspace. I wasn't thinking about fetching schemas over HTTP and at this point and I don't want to go there. There are complications involved in doing that right that we don't want to deal with right now. For this task, you should only be considering schemas defined in the workspace or pre-registered in `@hyperjump/json-schema` like the meta-schemas.
> 
> `my-project/schemas/address.schema.json`
> 
> ```json
> {
>   "$schema": "https://json-schema.org/draft/2020-12/schema",
>   "$id": "https://example.com/schemas/address",
> 
>   ...
> }
> ```
> 
> `my-project/schemas/customer.schema.json`
> 
> ```json
> {
>   "$schema": "https://json-schema.org/draft/2020-12/schema",
>   "$id": "https://example.com/schemas/customer",
> 
>   "type": "object",
>   "properties": {
>     ...,
>     "address": { "$ref": "/schemas/address" } <-- External reference to a schema in the workspace
>   }
> }
> ```

What exactly is workspace, is it just a folder containing many schema.json files ? If we open a folder in an IDE and the folder has many schema.json files and with them it has another folder which also has schema,json files then how will the workspaces be defined here?

File: language-server/src/references.js (hunk 3 above)
Comment (jdesrosiers) [reply language-server/src/references.js:75]: It's both. It's a reference to a location inside an external schema. You need to handle both it the schema doesn't exist and if the location doesn't exist within that external schema.

**Comment by jdesrosiers** [comment]
> What exactly is workspace

"Workspace" is an LSP concept. Open a directory in vscode and look at the file tree for that directory. That's your workspace. Everything that's visible to the IDE/editor is your workspace.

**Comment by probablyArth** [comment]
> > What exactly is workspace
> 
> "Workspace" is an LSP concept. Open a directory in vscode and look at the file tree for that directory. That's your workspace. Everything that's visible to the IDE/editor is your workspace.

ok! Just a little confused.

let's consider this example.

`my-project/schemas/b.schema.json`
```jsonc
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/c",

  ...
}
```

`my-project/schemas/c.schema.json`
```jsonc
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/b",

  ...
}
```

`my-project/schemas/a.schema.json`
```jsonc
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/a",
  "type": "object",
  "properties": {
    ...,
    "address": { "$ref": "/schemas/b" }
  }
  ...
}
```

is the ref in `my-project/schemas/a.schema.json` pointing to the file at `/schemas/b.schema.json` or `/schemas/c.schema.json` since `c.schema.json` has the id of `/schemas/b`

**Comment by jdesrosiers** [comment]
If you haven't read [Structuring a Complex Schema](https://json-schema.org/understanding-json-schema/structuring) on the website, definitely do that. It should help with the basic concepts.

There are two schema identification concepts. One is the retrieval URI and the other is self-identification.

Because these are files that exist on the file system, they have a `file:` URI that identifies them by their location on the file system. LSP uses this URI to identify files in the workspace. An example would be `file:///path/to/my-project/schemas/a.schema.json`.

However, JSON Schema also defines a way for schemas to declare their own identifier in addition to the natural identifier they get based on their file system location. This is done this using the `$id` keyword (`id` in draft-04). When a schema self-identifies, it effectively has two identifiers: the one it declares and it's file system location. LSP understands the natural identifier, but the self assigned identifier, so you need to implement that somehow.

So, going to your example,

LSP-ID: `file:///path/to/my-project/another-folder/b.schema.json`
Self-ID: `https://example.com/schemas/c`

{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/c",

  ...
}

LSP-ID: `file:///path/to/my-project/another-folder/c.schema.json`
Self-ID: `https://example.com/schemas/b`

{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/b",

  ...
}

LSP-ID: `file:///path/to/my-project/schemas/a.schema.json`
Self-ID: `https://example.com/schemas/a`

{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/a",
  "type": "object",
  "properties": {
    ...,
    "address": { "$ref": "/schemas/b" }
  }
  ...
}

The question is, what does the reference at `/properties/address` resolve to. `$id` doesn't just assign an identifier to the schema, it also sets the base URI. So, the URI-reference `/schemas/b` is resolved against the `https://example.com/schemas/a` to get `https://example.com/schemas/b`. There is only one schema identified by that exact URI and that's the one located at `/path/to/my-project/another-folder/c.schema.json`.

Now consider if the schema at `/path/to/my-project/schemas/a.schema.json` didn't self-identify (didn't have a `$id`). How would that change how the reference resolves? Let's also change the reference to `../another-folder/b.schema.json` to make the result interesting. Since there's no `$id`, the base URI is `file:///path/to/my-project/schemas/a/schema.json` and it resolves to `file://path/to/my-project/another-folder/b.schema.json`. This resolves the schema that self-identifies as `https://example.com/schemas/c`.

**Comment by probablyArth** [comment]
> If you haven't read [Structuring a Complex Schema](https://json-schema.org/understanding-json-schema/structuring) on the website, definitely do that. It should help with the basic concepts.
> 
> There are two schema identification concepts. One is the retrieval URI and the other is self-identification.
> 
> Because these are files that exist on the file system, they have a `file:` URI that identifies them by their location on the file system. LSP uses this URI to identify files in the workspace. An example would be `file:///path/to/my-project/schemas/a.schema.json`.
> 
> However, JSON Schema also defines a way for schemas to declare their own identifier in addition to the natural identifier they get based on their file system location. This is done this using the `$id` keyword (`id` in draft-04). When a schema self-identifies, it effectively has two identifiers: the one it declares and it's file system location. LSP understands the natural identifier, but the self assigned identifier, so you need to implement that somehow.
> 
> So, going to your example,
> 
> LSP-ID: `file:///path/to/my-project/another-folder/b.schema.json` Self-ID: `https://example.com/schemas/c`
> 
> ```json
> {
>   "$schema": "https://json-schema.org/draft/2020-12/schema",
>   "$id": "https://example.com/schemas/c",
> 
>   ...
> }
> ```
> 
> LSP-ID: `file:///path/to/my-project/another-folder/c.schema.json` Self-ID: `https://example.com/schemas/b`
> 
> ```json
> {
>   "$schema": "https://json-schema.org/draft/2020-12/schema",
>   "$id": "https://example.com/schemas/b",
> 
>   ...
> }
> ```
> 
> LSP-ID: `file:///path/to/my-project/schemas/a.schema.json` Self-ID: `https://example.com/schemas/a`
> 
> ```json
> {
>   "$schema": "https://json-schema.org/draft/2020-12/schema",
>   "$id": "https://example.com/schemas/a",
>   "type": "object",
>   "properties": {
>     ...,
>     "address": { "$ref": "/schemas/b" }
>   }
>   ...
> }
> ```
> 
> The question is, what does the reference at `/properties/address` resolve to. `$id` doesn't just assign an identifier to the schema, it also sets the base URI. So, the URI-reference `/schemas/b` is resolved against the `https://example.com/schemas/a` to get `https://example.com/schemas/b`. There is only one schema identified by that exact URI and that's the one located at `/path/to/my-project/another-folder/c.schema.json`.
> 
> Now consider if the schema at `/path/to/my-project/schemas/a.schema.json` didn't self-identify (didn't have a `$id`). How would that change how the reference resolves? Let's also change the reference to `../another-folder/b.schema.json` to make the result interesting. Since there's no `$id`, the base URI is `file:///path/to/my-project/schemas/a/schema.json` and it resolves to `file://path/to/my-project/another-folder/b.schema.json`. This resolves the schema that self-identifies as `https://example.com/schemas/c`.

got it! I almost wrote another huge example, but I read the comment once again and now I completely get it!
Thank you so much, I will finish this ASAP!

**Comment by probablyArth** [comment]
I think I'm done with all the tasks and my PR is ready for review!!

**Comment by probablyArth** [comment]
https://github.com/hyperjump-io/json-schema-language-tools/assets/61962871/76f57e5b-a54d-43d4-952c-4c9d51c501c8

https://github.com/hyperjump-io/json-schema-language-tools/assets/61962871/2193cbcf-c4da-4f36-a658-605862b80dfb

Here are the demo videos! @jdesrosiers

**Comment by jdesrosiers** [comment]
Please rebase and resolve conflicts. Also, make sure you run tests locally to make sure they pass before pushing. Don't rely on automation.

**Comment by probablyArth** [comment]
> Please rebase and resolve conflicts. Also, make sure you run tests locally to make sure they pass before pushing. Don't rely on automation.

The failed checks don't seem to be related the changes I made. Could you please take a look?
image

**Comment by jdesrosiers** [comment]
It's definitely related to your changes. The tests pass on "main".

File: language-server/src/util.js [hunk 4]
```diff
@@ -1 +1,50 @@
+import { DiagnosticSeverity } from "vscode-languageserver";
+import { documents } from "./server.js";
```
Comment (jdesrosiers) [inline language-server/src/util.js:2]: Here's your problem

**Comment by probablyArth** [comment]
Thanks for pointing that out @jdesrosiers 
I made the necessary changes, tests should pass now.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
A found a couple bugs in some edge cases you probably hadn't considered.

**Reference to an embedded Schema**

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/main",

  "allOf": [
    { "$ref": "a" },
    { "$ref": "b" },
  ]

  "$defs": {
    "embedded-a": {
      "$id": "a"
    },
    "embedded-b": {
      "$id": "https://example.com/b"
    }
  }
}
```

These are valid references that get flagged as invalid.

**$ref in a non-schema location**

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/main",

  "properties": {
    "$ref": "foo"
  },
  "not-a-keyword": { "$ref": "foo" }
}
```

These get flagged as invalid references, but neither are references. They just look like references.

---

The design isn't great in that finding references is coupled to validating references. There are other features where we'll need to use or know about references. Ideally, we'd be able to reuse the reference identification functionality you've developed for those other features, but we can't because it's too coupled to validating references.

Similarly, I'd like to see the validation of the references decoupled from language server concerns. Eventually, I want to use the functionality developed here to make a CLI that validates all the schemas in a workspace. That means I need to use this validation functionality outside of the context of a language server. So, things like calling `buildDiagnostic` is a problem. It should just return the validation information. Then we can call it from the language server and run `buildDiagnostic` on the result or call it from the CLI and run whatever is need to present those errors on the console.

**Comment by probablyArth** [comment]
- Got it. Yes I forgot about those cases.
- That's exactly what I was thinking a couple hours ago when I stumbled upon #17 , I also feel the same. I will definitely decouple the logic.

I will get on that asap!

File: language-server/src/references.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    "https://json-schema.org/keyword/ref",
+    "https://json-schema.org/keyword/draft-04/ref"
+  ];
+  const referenceKeywordNames = referenceKeywordIds.map((keywordId) => keywordNameFor(keywordId, dialectUri));
+  if (!referenceKeywordNames.includes(keywordInstance.value()) || keywordInstance?.node?.parent?.children?.length < 2) {
+    return;
+  }
+  const { pointer, textDocument, node } = keywordInstance;
+
+  const valueInstance = keywordInstance._fromNode(node.parent.children[1], keywordInstance.pointer);
```
Comment (jdesrosiers) [inline language-server/src/references.js:43]: The point of the `JsoncInstance` class is to abstract the underlying AST being used. We're currently using `jsonc-parser`'s AST, but that will probably need to change at some point and ideally, all we should have to do is to replace `JsoncInstance` with a new `*Instance` class and everything should still work.

All that is to say, don't use `node` outside of `JsoncInstance`. I should have made it private, but support for private members in classes is a relatively new feature and I forgot it existed. If you need to traverse the AST, use the utility functions provide in `JsoncInstance` or make a new one if what you need isn't available.

Also, `_fromNode` was intended to be private as well.

**Review by jdesrosiers — COMMENTED** [review]
I haven't gone through this in detail yet, but this is a lot more code than I was expecting. I feel like this should be simpler. I think you're duplicating things that are already handled. I would expect you to make use of the managed instances. Every time the workspace changes, every schema resource in the workspace is revalidated and is available in the `schemaResourceCache`. The schemas are already decomposed so there are no embedded schemas to worry about. You just need to collect the identifiers for every schema resource to get the list of valid identifiers. Then you just need to iterate over those schemaResources looking for and validating references, which shouldn't be too hard since you don't have to worry about embedded schemas.

**Comment by probablyArth** [comment]
Hey @jdesrosiers , I have refactored my solution more.

I have removed the identifiers store but I have kept the `inactiveDocumentStore` , since that caches the documents which are not  open in the editor and saves a lot of io operations. and the instances are removed from the `inactiveDocumentStore` once those are open in the editor.

File: language-server/src/server.js [hunk 6]
```diff
@@ -247,13 +251,26 @@ const validateSchema = async (textDocument) => {
...
       }
-
       continue;
     }
 
     try {
       const annotatedInstance = await annotate(dialectUri, schemaInstance);
 
+      const references = [];
+      for (const { keywordInstance } of getSemanticTokens(schemaResources)) {
```
Comment (jdesrosiers) [inline language-server/src/server.js:261]: It's surprising to see `setSemanticTokens` here. I think I see what you're doing though. It's a hack to get locations that are keywords. I suppose it's fine for now because we don't have a better way, but please leave a comment explaining why this is used here.

File: language-server/src/server.js [hunk 7]
```diff
@@ -247,13 +251,26 @@ const validateSchema = async (textDocument) => {
...
+        const found = findRef(keywordInstance, dialectUri);
+        if (typeof found !== "undefined") {
+          references.push(found);
+        }
+      }
+      for (const ref of references) {
+        const output = await validateReference(documents, textDocument, ref);
+        if (!output.valid) {
+          const instance = JsoncInstance.fromTextDocument(textDocument).get("#" + ref);
```
Comment (jdesrosiers) [inline language-server/src/server.js:270]: Do you really need to construct a whole new instance here? Can you just do `schemaResource.get`?

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
> I have kept the `inactiveDocumentStore`, since that caches the documents which are not open in the editor

I think you've misunderstood what I was telling you in my last review. The `schemaResourceCache` should contain _all_ schemas in the workspace, not just the ones that are open. All of them. You shouldn't need to `invalidDocumentStore`. It should all be there already.

File: language-server/src/server.js [hunk 8]
```diff
@@ -31,24 +29,29 @@ import { JsoncInstance } from "./jsonc-instance.js";
...
 import { getSemanticTokens } from "./semantic-tokens.js";
+import { buildDiagnostic, isSchema } from "./util.js";
+import { deleteFromInactiveDocumentStore, fetchDocument } from "./documents.js";
+import { findRef, validateReference } from "./references.js";
 
 
 setMetaSchemaOutputFormat(DETAILED);
 setShouldValidateSchema(false);
 
-const isSchema = RegExp.prototype.test.bind(/(?:\.|\/|^)schema\.json$/);
```
[the remark this answers is not in the document]
Comment (probablyArth) [reply language-server/src/server.js:39]: :+1:

**Comment by probablyArth** [comment]
I have made some changes, please take a look @jdesrosiers !

**Comment by probablyArth** [comment]
Seems like there are a lot of changes made for schema resources :sweat_smile: 
I'll update my PR or maybe open a new one at this point :sweat_smile:

**Comment by jdesrosiers** [comment]
Yes, sorry. There was a big change introducing a more powerful and performant AST for working with schemas that replaced `JsoncInstance`. The change is preparing for supporting more advanced features like linting.

**Comment by jdesrosiers** [comment]
Closing in favor of #44
